opentelemetry-sdk/src/opentelemetry/sdk/error_handler/__init__.py

```diff
@@ -105,26 +106,26 @@ def __new__(cls) -> "GlobalErrorHandler":
 
         return cls._instance
 
-    def __enter__(self):
+    def __enter__(self) -> None:
         pass
 
     # pylint: disable=no-self-use
-    def __exit__(self, exc_type, exc_value, traceback):
-        if exc_value is None:
+    def __exit__(self, exc_type, exc_value, traceback) -> Optional[bool]:  # type: ignore
+        if exc_value is None:  # type: ignore
             return None
 
         plugin_handled = False
 
-        error_handler_entry_points = entry_points(
+        error_handler_entry_points = entry_points(  # type: ignore
```

Comment:
This can be typed, https://github.com/python/importlib_metadata/blob/main/importlib_metadata/__init__.py#L1039C42-L1039C43
Need to check if we are exporting it from our util module

Reply:
It is - https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-api/src/opentelemetry/util/_importlib_metadata.py#L28

Reply:
basedpyright: Import "opentelemetry.util._importlib_metadata" could not be resolved. This results in following errors when mypy is executed - 

opentelemetry-sdk/src/opentelemetry/sdk/error_handler/__init__.py:119: error: Type of variable becomes "Any" due to an unfollowed import  [no-any-unimported]
opentelemetry-sdk/src/opentelemetry/sdk/error_handler/__init__.py:119: error: Expression has type "Any"  [misc]

Reply:
@xrmx Please let me know if the above errors can be handled differently.

Reply:
I mean something like this `error_handler_entry_points : Entrypoints = entry_points`

Reply:
> I mean something like this `error_handler_entry_points : Entrypoints = entry_points(`

Sorry, I should have added the changes that resulted in the above error.

These errors 
`opentelemetry-sdk/src/opentelemetry/sdk/error_handler/init.py:119: error: Type of variable becomes "Any" due to an unfollowed import [no-any-unimported]
opentelemetry-sdk/src/opentelemetry/sdk/error_handler/init.py:119: error: Expression has type "Any" [misc]` are happening after adding the type `error_handler_entry_points : Entrypoints = entry_points(`

Reply:
@xrmx please let me know if adding `_importlib_metadata.pyi` will be a good idea to resolve this issue.

Reply:
> error_handler_entry_points : Entrypoints = entry_points(

This is a workaround, `entry_points` is a `Entrypoints`, why would you need to add a type to it? The type checker should be able to infer. If you have to do this, it's a sign that something is wrong.

 > @xrmx please let me know if adding _importlib_metadata.pyi will be a good idea to resolve this issue.
 
 Adding `pyi` files nowadays don't make sense. The `.pyi` files are useful to projects that create python objects in other languages, like rust, or C code, and they want to provide an interface for type checkers to understand those types.

---

The error mentioned here:

> basedpyright: Import "opentelemetry.util._importlib_metadata" could not be resolved. This results in following errors when mypy is executed -
> opentelemetry-sdk/src/opentelemetry/sdk/error_handler/init.py:119: error: Type of variable becomes "Any" due > to an unfollowed import [no-any-unimported]
> opentelemetry-sdk/src/opentelemetry/sdk/error_handler/init.py:119: error: Expression has type "Any" [misc]

It's because `opentelemetry.util` is not found. I have the feeling that the setup here is wrong, and there have been happening some workarounds like the suggested to make things "work" (bypass the issue).

Reply:
Anyway, what should be done here is to make the tool find the `opentelemetry.util`... Maybe something wrong with the `MYPY_PATH` on the tox file? I'm not sure...

Reply:
@Kludex, thank you. When you say setup, do you mean setup on my local machine or project setup in general? If it is project setup, do you have any suggestions on how do I go about rectifying it? Thank you.

Reply:
> @Kludex, thank you. When you say setup, do you mean setup on my local machine or project setup in general?

Given that the project runs tox, you shouldn't have a different output from your local machine, and the CI. So... I meant the project setup.

I guess that you are in `opentelemetry/sdk/`, and the other path is `opentelemetry/util/`, so you have the same root, and mypy gets confused. I'm not sure how to tweak the configuration to make this work. I've tried to reproduce the analogous settings on `opentelemetry-contrib-python`, and I thought I was doing something wrong, but this PR makes me think that the setup here is missing something.

Anyway, I don't have a solution.

Reply:
I tried your branch... And it doesn't seem mypy is running in this file... How did you run it? 🤔

Reply:
> I tried your branch... And it doesn't seem mypy is running in this file... How did you run it? 🤔

We have to add `mypy: mypy --install-types --non-interactive --namespace-packages --explicit-package-bases opentelemetry-sdk/src` to `tox.ini`.
I haven't included it in the review as this is a partial change.

Reply:
Ah... You are a first time contributor. The pipeline is not running...

Reply:
Ah ok. I added:
```
mypy: mypy --install-types --non-interactive --namespace-packages --explicit-package-bases opentelemetry-sdk/src/opentelemetry/sdk/error_handler
```

I've confirmed. `mypy` cannot find `opentelemetry/util`, so it converts `entry_points` to `Any`.

Reply:
`mypy: MYPYPATH={toxinidir}/opentelemetry-api/src/:{toxinidir}/opentelemetry-semantic-conventions/src/:{toxinidir}/opentelemetry-sdk/src/:{toxinidir}/tests/opentelemetry-test-utils/src/`

includes `{toxinidir}/opentelemetry-api/src/`.

Reply:
Yeah, so my point is that it should have been found, but it's not.

Reply:
When I changed 
`from importlib_metadata import (` 
to 
`from importlib.metadata import (`

mypy is able to find the imports. `importlib_metadata` is valid for python < 3.8.

Reply:
Will update the PR to change `importlib_metadata` to `imortlib.metadata` and will remove `ignore` in the files - 
	`opentelemetry-api/src/opentelemetry/context/__init__.
        opentelemetry-api/src/opentelemetry/propagate/__init__.py
	opentelemetry-api/src/opentelemetry/util/_importlib_metadata.py
	 opentelemetry-api/src/opentelemetry/util/_providers.py`

Reply:
@xrmx / @Kludex, Can this be suppressed with `# type: ignore` for the time.

Reply:
I don't think `type: ignore` should be used in any situation. It's usually a sign that something is wrong. **At least** the reason to ignore should be mentioned in the brackets. That's why I suggested this: https://github.com/open-telemetry/opentelemetry-python/pull/4360#pullrequestreview-2511607423

I think the PR I opened should help you: https://github.com/open-telemetry/opentelemetry-python/pull/4368 (?)

Reply:
@Kludex updated the PR as suggested. Let me know if I can resolve this conversation.

Reply:
This PR has more type ignores than actual typing improvements.